### ISSUE TITLE
Mask structure question - with visualization

### ISSUE THREAD

**Issue by Wonder1905** [allenzren/open-pi-zero#4]
Hey, thanks for our hard and amazing work!
I have a small question regarding the masking.
Im runinng:
```
python scripts/try_checkpoint_in_simpler.py     --task google_robot_pick_horizontal_coke_can     --checkpoint_path fractal_beta_step29576_2024-12-29_13-10_42.pt     --recording     --use_bf16     --use_torch_compile # first batch will be slow
```
And trying to visualize the mask, from line 113:
```        inputs = env_adapter.preprocess(env, obs, instruction)
        causal_mask, vlm_position_ids, proprio_position_ids, action_position_ids = (
            model.build_causal_mask_and_position_ids(
                inputs["attention_mask"], dtype=dtype
            )
        )
        visualize_causal_mask(causal_mask,name='causal_mask.png')
        image_text_proprio_mask, action_mask = model.split_full_mask_into_submasks(
            causal_mask
        )
```
Where the visualization function is just a simple one:
```
def visualize_causal_mask(causal_mask,name='causal_mask.png'):
    causal_mask = causal_mask.squeeze().float()
    if len(causal_mask.shape) == 1:
        causal_mask = causal_mask[:, None] * causal_mask[None, :]
    # Normalize the mask values between 0 and 1 for visualization
    causal_mask = (causal_mask - causal_mask.min()) / (causal_mask.max() - causal_mask.min())
    plt.figure(figsize=(10,10))
    plt.imshow(causal_mask.cpu().numpy(), cmap='binary')
    plt.colorbar()
    plt.title('Causal Mask')
    plt.xlabel('Sequence Position (j)')
    plt.ylabel('Sequence Position (i)') 
    plt.savefig(name)
    plt.close()
```

The mask does not seems as expected and I will be happy if you can help me understand (the title is just a name I know that the mask is not suppose to be just casual):
![causal_mask](https://github.com/user-attachments/assets/bff6691c-0e39-4b60-ba13-bdd1e0416585)

Thanks a lot !

**Comment by allenzren** [comment]
I think this is correct? If you zoom in, there is a small block for the proprio token.

It is 276 (256 image + max 20 text, so some white padding) + 1 proprio + 4 actions right now

**Comment by allenzren** [comment]
Basically, you can imagine the top left block here is much much bigger, then they look the same?

https://github.com/allenzren/open-pi-zero/blob/main/media/open-pi-zero-overview.png

**Comment by Wonder1905** [comment]
I think whats bothering me is the non causality  of the mask in the text and image part. Even in the "prefill phase" there should be a causal mask. At least in my humble opinion. But maybe Im missing something

**Comment by allenzren** [comment]
Ah I see. I don't think the original Pi0 used such casual masking within the VLM part though. This is taken from Appendix B
<img src="https://github.com/user-attachments/assets/a4c51182-ac12-4543-9ebf-ce1a4ab618cc" width="480">

It says the blocks are full bidirectional, and the casual part is only block-wise. Since they didn't train the model with text data, I don't think they needed any casual masking for img/text?

**Comment by Wonder1905** [comment]
You are right it is clearly written, sorry for my confusion.
